src/ra_log_segment.erl

```diff
@@ -258,10 +261,65 @@ read(State, Idx, Num) ->
 -spec read_cons(state(), ra_index(), Num :: non_neg_integer(),
                 fun((binary()) -> term()), Acc) ->
     Acc when Acc :: [{ra_index(), ra_term(), binary()}].
-read_cons(#state{cfg = #cfg{fd = _Fd, mode = read},
-                 index = Index} = State, Idx,
-          Num, Fun, Acc) ->
-    pread_cons(State, Idx, Idx + Num - 1, Index, Fun, Acc).
+read_cons(#state{cfg = #cfg{mode = read} = Cfg,
+                 cache = Cache,
+                 index = Index}, Idx, Num, Fun, Acc) ->
+    pread_cons(Cfg, Cache, Idx, Idx + Num - 1, Index, Fun, Acc).
+
+read_sparse(#state{index = Index,
+                   cfg = Cfg,
+                   cache = _Cache0}, Indexes, Fun, Acc) ->
+    Cache0 = prepare_cache(Cfg, Indexes, Index),
+    Entries = read_sparse0(Cfg, Indexes, Index, Cache0, Fun, Acc),
+    {undefined, length(Entries), Entries}.
+
+read_sparse0(_Cfg, [], _Index, _Cache, _Fun, Acc) ->
+    Acc;
+read_sparse0(Cfg, [NextIdx | Rem] = Indexes, Index, Cache0, Fun, Acc) ->
+    {Term, Offset, Length, _} = map_get(NextIdx, Index),
+    case cache_read(Cache0, Offset, Length) of
+        false ->
+            case prepare_cache(Cfg, Indexes, Index) of
+                undefined ->
+                    {ok, Data, _} = pread(Cfg, undefined, Offset, Length),
+                    read_sparse0(Cfg, Rem, Index, undefined, Fun,
+                                 [{NextIdx, Term, Fun(Data)} | Acc]);
+                Cache ->
+                    read_sparse0(Cfg, Indexes, Index, Cache, Fun, Acc)
+            end;
+        Data ->
+            read_sparse0(Cfg, Rem, Index, Cache0, Fun,
+                         [{NextIdx, Term, Fun(Data)} | Acc])
+    end.
+
+cache_read({CPos, CLen, Bin}, Pos, Length)
+  when Pos >= CPos andalso
+       Pos + Length =< (CPos + CLen) ->
+    %% read fits inside cache
+    binary:part(Bin, Pos - CPos, Length);
+cache_read(_, _, _) ->
+    false.
+
+prepare_cache(#cfg{} = _Cfg, [_], _SegIndex) ->
+    undefined;
+prepare_cache(#cfg{fd = Fd} = _Cfg, [FirstIdx | Rem], SegIndex) ->
+    case consec_run(FirstIdx, FirstIdx, Rem) of
+        {Idx, Idx} ->
+            %% no run, no cache;
+            undefined;
+        {FirstIdx, LastIdx} ->
+            {_, FstPos, _FstLength, _} = map_get(FirstIdx, SegIndex),
+            {_, LastPos, LastLength, _} = map_get(LastIdx, SegIndex),
+            %% The cache needs to be at least as large as the next entry
+            %% but no larger than ?READ_AHEAD_B
+            MaxCacheLen = LastPos + LastLength - FstPos,
+            %% read at least the remainder of the block from
+            %% the first position
+            MinCacheLen = ?BLOCK_SIZE - (FstPos rem ?BLOCK_SIZE),
+            CacheLen = max(MinCacheLen, min(MaxCacheLen, ?READ_AHEAD_B)),
```

Comment:
What if `FstLength` is greater than `?READ_AHEAD_B`, but lower than `MinCacheLen`? Like `FstPos` is close to the end of the block. I guess you still want to use `FstLength` as the minimum somewhere in the computation.